Feature/visualisation api

istefano82 opened the pull request:
Learning curves visualisation API.

 - Split loading of visualisation data and displaying it in different functions.
 - Adapt argparser to first load the data than pass it to the function for visualisation,
 - Create utility functions to reuse code for path manipulation and field ground truth validation.
- Start API integration tests for saving PNG and PDF viz formats.

Commit:
Added file output for visualization

Commit:
Fix compare_classifiers_multiclass_multimetric

Commit:
Additional check on the presence of per_class_stats in test statistics

Commit:
Change in plotting: shows the plot only if an output dir is not specified

Commit:
Fixed confidence_thresholding_data_vs_acc_subset_per_class

Commit:
Fix #346: added a parameter indicating if the session should be closed after training in full_train

Commit:
Fix #351: values in categorical columns are stripped before being compared to the vocabulary

w4nderlust committed:
Improved Ludwig logo print

w4nderlust committed:
Merge remote-tracking branch 'origin/master'

Commit:
Fix #354: replace --horovod with --use_horovod in the user guide

Commit:
Fix #345: replaced description for the save function in api

Commit:
Removed unused code

istefano82 committed:
Fix issues causing the visualisation tests to fail

istefano82 committed:
Fix issues causing the visualisation tests to fail

istefano82 committed:
Visualisations integration test suite.

istefano82 committed:
Visualisations integration test suite.

msaisumanth committed:
Refactors the sequence encoders code and adds some unittests (#342)

* Refactors the sequence encoders code and adds some unittests. Removes a duplicate test file

* Marks a missing TODO

msaisumanth committed:
Fix #340: Configures logging in ludwig/__init__ (#352)

* Configures logging in ludwig/__init__

* Modifies the usage of logging to support a central configuration for the project

danicattaneob committed:
#357 Added experiment name parameter (#359)

Commit:
Updated docs with new parameters and fixed typos

istefano82 committed:
SPIKE - learning curves api visualisation and integration test.

istefano82 committed:
SPIKE - learning curves api visualisation and integration test.

Commit:
Fix #364: associate the right function to non english text format functions

istefano82 committed:
Learning curves visualisation now expects processed training data as list.

1. Extract loading of data for visualisation in to separate function to be called
by the argparser and than pass the loaded data to the visualisation function.
2. Create function for data manipulation that extact prediction fields from training data.

istefano82 committed:
Learning curves visualisation now expects processed training data as list.

1. Extract loading of data for visualisation in to separate function to be called
by the argparser and than pass the loaded data to the visualisation function.
2. Create function for data manipulation that extact prediction fields from training data.

Commit:
Fixed edit uri in documentation

Commit:
Fix #372: set evaluate performance parameter to false in predict.py

istefano82 committed:
Refactor learning curves visualisation and helper methods for loading and processing data.

Adapt test for PNG and PDF formats.

istefano82 committed:
Refactor learning curves visualisation and helper methods for loading and processing data.

Adapt test for PNG and PDF formats.

gogasca committed:
Update as_matrix() -> values in numerical and binary features (#373)

istefano82 committed:
Create function to generate filane template path and create output directory.

istefano82 committed:
Create function to generate filane template path and create output directory.

w4nderlust commented on ludwig/visualize.py:
```diff
@@ -2233,6 +2276,8 @@ def cli(sys_argv):
         frequency_vs_f1(**vars(args))
     elif args.visualization == 'learning_curves':
+        training_statistics_per_model_name = load_data_for_viz(**vars(args))
+        setattr(args, 'training_statistics_per_model_name', training_statistics_per_model_name)
```
I would actually pass the `training_statistics_per_model_name` to `learning_curves()` like:
```
learning_curves(training_statistics_per_model_name= training_statistics_per_model_name, **vars(args))
```
because args contains `training_statistics` and not `training_statistics_per_model_name` so the two names don't clash.

istefano82 replied:
I agree it makes sense to pass it as a keyword argument instead of setting as atribute of the args object.

w4nderlust commented on ludwig/visualize.py:
```diff
@@ -36,45 +36,88 @@
-):
-    if len(training_statistics) < 1:
-        logging.error('No training_statistics provided')
+def load_data_for_viz(training_statistics,**kwargs):
```
this function will eventually load everything, also probs, preds, etc, right?

istefano82 replied:
I plan to do so. Maybe it will be an handler function that will call specific loaders for each type of data for visualisation and return it to the argparser before it is been sent to the visualisation function. 

I expect to have more clear design once I start adapting the rest of the visualisation functions.

w4nderlust replied:
Makes sense. Probably trying to load everything and returning `None` for the missing ones and then letting the code in `cli()` pass the loaded files it needs to the actual visualization function is a good idea to start with, but yeah, you'll figure out by doing it.

w4nderlust commented on ludwig/visualize.py:
```diff
@@ -36,45 +36,88 @@
+        return os.path.join(output_dir, filename_template)
+    return None
+
+def learning_curves(
```
this looks fine for learning curves, I would just:
- rename `training_statistics_per_model_name` to something shorter, maybe `training_statistics_per_model`
- make it work for both lists and single values, like check if it is a list, if not do `training_statistics_per_model = [training_statistics_per_model]`
- the same thing applies to `model_names`, make it work both if it is a string or a list of strings
- the same for `field`, make it work for one single string or a list of strings

istefano82 replied:
Type checking is not very pythonic paradigm - “it’s easier to ask for forgiveness than permission”. But maybe I can write converter functions that do so 

Or we can ask  API users to pass them as iterable (list or tuple) so that the functionality continues to work - I am currently writing the integration tests in such way.

Let me know which one you prefer.

w4nderlust replied:
From the developer experience point of view, if I have only one training statistics file and one algorithm name, I would prefer to call `learning_curves(training_stat, alg_name)` rather than `learning_curves([training_stat], [alg_name])`.
If there are better ways than checking for the type, that's great, you can figure it out :)

w4nderlust commented:
This is a great starting point! Left a few comments, but overall it's great.

istefano82 committed:
Create to list convertor and use it on learning curves field and model_names arguments.

- Put back argparser default value for the field to empty list.
- Pass the extracted training data per model to the visualisation function.

istefano82 committed:
Create to list convertor and use it on learning curves field and model_names arguments.

- Put back argparser default value for the field to empty list.
- Pass the extracted training data per model to the visualisation function.

istefano82 committed:
Make visualisation data loader to generic - i.e. loading(train, test predict data).

istefano82 committed:
Make visualisation data loader to generic - i.e. loading(train, test predict data).

istefano82 committed:
Make load data for viz more generic.

Now expects explicit path to the statistics as argument.

istefano82 committed:
Make load data for viz more generic.

Now expects explicit path to the statistics as argument.

istefano82 committed:
It is now possible to call compare perfomance visualisation programatically via test statistics.

Split data loading processing visualisation logic of compare perfomance visualisation.
Write integration API test for PDF and PNG outputs.

istefano82 committed:
It is now possible to call compare perfomance visualisation programatically via test statistics.

Split data loading processing visualisation logic of compare perfomance visualisation.
Write integration API test for PDF and PNG outputs.

kvthr committed:
Added normalization strategies for numerical features (#367)

istefano82 committed:
SPIKE - create compare_classifiers_performance_from_prob vis API and test.

CLI test now fails.

istefano82 committed:
SPIKE - create compare_classifiers_performance_from_prob vis API and test.

CLI test now fails.

Commit:
Updated sequence tagging example in examples.md

Commit:
Added examples of text classification datasets to exmaples.md

Commit:
Website update

istefano82 committed:
Adapted compare_classifiers_performance_from_prob CLI and integration test.

istefano82 committed:
Adapted compare_classifiers_performance_from_prob CLI and integration test.

istefano82 committed:
Remove the unused field parameter from the compare perfomance from prob.

Adapt integration API test.

istefano82 committed:
Remove the unused field parameter from the compare perfomance from prob.

Adapt integration API test.

Commit:
Create API test for compare performance from pred.

Commit:
Create API test for compare performance from pred.

istefano82 committed:
Adapted compare_classifiers_performance_from_pred to separate loading of data from manipulating and visualising.

API test fails. CLi test passes.

istefano82 committed:
Adapted compare_classifiers_performance_from_pred to separate loading of data from manipulating and visualising.

API test fails. CLi test passes.

Commit:
Fix #377: correct favicon in the non-landing pages

IgorWilbert committed:
Fixes style in landing page footer (#378)

Commit:
Updated docs

Commit:
Fix: fixed resizing of images

Commit:
from info to debug when logging hidden shape

ydudin3-zz committed:
Refactor text formatters (#376)

Commit:
BugFix parameter s -> text in tokenizers

istefano82 committed:
Add todo to update load from file parameters to be able to select split type.

istefano82 committed:
Add todo to update load from file parameters to be able to select split type.

istefano82 committed:
Fix wrong calculation of ground truth data for compare classifiers from prob test.

istefano82 committed:
Fix wrong calculation of ground truth data for compare classifiers from prob test.

istefano82 committed:
Extract data csv split in to helper function.

istefano82 committed:
Extract data csv split in to helper function.

istefano82 committed:
Fix wrong calculation of ground truth data for compare classifiers from pred test.

istefano82 committed:
Fix wrong calculation of ground truth data for compare classifiers from pred test.

istefano82 committed:
Adapted compare_classifiers_performance_subset to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted compare_classifiers_performance_subset to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Convert ground truth lists to numpy arrays.

Cleanup redundant  arguments passed to mode.train.

istefano82 committed:
Convert ground truth lists to numpy arrays.

Cleanup redundant  arguments passed to mode.train.

istefano82 committed:
FIx Typo.

istefano82 committed:
FIx Typo.

ydudin3-zz committed:
Add BERT tokenizer (#380)

Commit:
Fix embedding size in parallel cnn encoder

w4nderlust commented on ludwig/utils/data_utils.py:
```diff
@@ -289,13 +289,16 @@ def class_counts(dataset, labels_field):
         column = hdf5_data[field].value
         hdf5_data.close()
+        # @TODO: use the split type column[split == split_type]
         array = column[split == 2]  # ground truth
```
as per our discussion about assumptions, you can remove this line entirely (which also mean we don't need to add an additional parameter. Obviously this will impact everywhere we use this function to load data, but will make things much cleaner

istefano82 replied:
I am not sure what you mean. Maybe we can discuss online.

istefano82 committed:
Adapted compare_classifiers_performance_schanging_k to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted compare_classifiers_performance_schanging_k to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Chage keyword argument passing to positional.

Vis function declaration do not require first arg to be keyword.

istefano82 committed:
Chage keyword argument passing to positional.

Vis function declaration do not require first arg to be keyword.

istefano82 committed:
Adapted compare_classifiers_multiclass_multimetric to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted compare_classifiers_multiclass_multimetric to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
convert all model_name arguments to list.

istefano82 committed:
convert all model_name arguments to list.

istefano82 committed:
convert all model_name arguments to list.

istefano82 committed:
convert all model_name arguments to list.

patrickvonplaten committed:
Update one-shot siamese network example (#382)

witdth and height should be under preprocessing

IgorWilbert committed:
Fixes footer style in application.css and runs mkdocs build (#381)

Commit:
Minor change in examples.md

Commit:
Update website

istefano82 committed:
Adapted compare_classifiers_prediction_distribution to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.
Integration tests are unstable sometime fail with:

File "/home/ivo-pc/Projects/ludwig/ludwig/visualize.py", line 786, in compare_classifiers_predictions_distribution
    filename=filename
  File "/home/ivo-pc/Projects/ludwig/ludwig/utils/visualization_utils.py", line 269, in radar_chart
    predictions = [pred[gt_argsort] for pred in predictions]
  File "/home/ivo-pc/Projects/ludwig/ludwig/utils/visualization_utils.py", line 269, in <listcomp>
    predictions = [pred[gt_argsort] for pred in predictions]
IndexError: index 2 is out of bounds for axis 0 with size 2

istefano82 committed:
Adapted compare_classifiers_prediction_distribution to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.
Integration tests are unstable sometime fail with:

File "/home/ivo-pc/Projects/ludwig/ludwig/visualize.py", line 786, in compare_classifiers_predictions_distribution
    filename=filename
  File "/home/ivo-pc/Projects/ludwig/ludwig/utils/visualization_utils.py", line 269, in radar_chart
    predictions = [pred[gt_argsort] for pred in predictions]
  File "/home/ivo-pc/Projects/ludwig/ludwig/utils/visualization_utils.py", line 269, in <listcomp>
    predictions = [pred[gt_argsort] for pred in predictions]
IndexError: index 2 is out of bounds for axis 0 with size 2

patrickvonplaten committed:
update siamese network example (#383)

patrickvonplaten committed:
Changed encoder description cnn_rnn to cnnrnn (#385)

istefano82 committed:
Adapted confidence_treshholding visualisation to be used from API and CLI.

Mdified integration API testcase to correctly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_treshholding visualisation to be used from API and CLI.

Mdified integration API testcase to correctly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_data_vs_acc to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_data_vs_acc to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_data_vs_acc_subset to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_data_vs_acc_subset to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_data_vs_acc_subset_per_class to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_data_vs_acc_subset_per_class to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_2thresholds_2d to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_2thresholds_2d to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_2thresholds_3d to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted confidence_thresholding_2thresholds_3d to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Simplify output features for treshhold 2d and 3d visualisation.

Correcly pass only single model and both output features probabilities.

istefano82 committed:
Simplify output features for treshhold 2d and 3d visualisation.

Correcly pass only single model and both output features probabilities.

istefano82 committed:
Adapted binary_threshold_vs_metric to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted binary_threshold_vs_metric to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted roc_curves to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted roc_curves to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted roc_curves_from_test_statistics to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted roc_curves_from_test_statistics to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted calibration_1_vs_all to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted calibration_1_vs_all to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted binary_calibration_multiclass to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted binary_calibration_multiclass to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted onfusion_matrix to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted onfusion_matrix to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted frequency_vs_f1 to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Adapted frequency_vs_f1 to be used from API and CLI.

Mdified integration API testcase to correcltly compute ground truth data and pass correct number of top n classes.

istefano82 committed:
Remove stdout and err pipelining from most of the tests.

istefano82 committed:
Remove stdout and err pipelining from most of the tests.

Commit:
add rquired normalize parameter to the confusion matrix test.

Commit:
add rquired normalize parameter to the confusion matrix test.

Commit:
Begin using the Experiment class in tests to reduce code duplication.

Commit:
Begin using the Experiment class in tests to reduce code duplication.

istefano82 committed:
Use Experiment class where possible to reduce code duplication.

istefano82 committed:
Use Experiment class where possible to reduce code duplication.

istefano82 committed:
Reformat to ensure PEP8 standarts are met.

istefano82 committed:
Reformat to ensure PEP8 standarts are met.

istefano82 committed:
Change method name.

istefano82 committed:
Change method name.

istefano82 committed:
Create compare_performance_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_performance_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create learning_cuves_cli wrapper function that loads model data for visualisation.

istefano82 committed:
Create learning_cuves_cli wrapper function that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_from_prob_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_from_prob_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
remove stdour stderr pipeline.

istefano82 committed:
remove stdour stderr pipeline.

istefano82 committed:
Create compare_classifiers_performance_from_pred_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_from_pred_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_from_subset_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_from_subset_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_changing_k_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Create compare_classifiers_performance_changing_k_cli wrapper funciton that loads model data for visualisation.

istefano82 committed:
Remove stdout/stderr from more tests

istefano82 committed:
Remove stdout/stderr from more tests

istefano82 committed:
Remove stdout/stderr from more tests

istefano82 committed:
Remove stdout/stderr from more tests

Commit:
Create compare_classifiers_performance_changing_k_cli wrapper funciton that loads model data for visualisation.

Commit:
Create compare_classifiers_performance_changing_k_cli wrapper funciton that loads model data for visualisation.

Commit:
Create compare_classifiers_predictions_cli wrapper funciton that loads model data for visualisation.

Commit:
Create compare_classifiers_predictions_cli wrapper funciton that loads model data for visualisation.

Commit:
Create compare_classifiers_predictions_distribution_cli wrapper funciton that loads model data for visualisation.

Commit:
Create compare_classifiers_predictions_distribution_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_data_vs_acc_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_data_vs_acc_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_data_vs_acc_subset_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_data_vs_acc_subset_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_data_vs_acc_subset_per_class_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_data_vs_acc_subset_per_class_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_2thresholds_2d_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_2thresholds_2d_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_2thresholds_3d_cli wrapper funciton that loads model data for visualisation.

Commit:
Create confidence_thresholding_2thresholds_3d_cli wrapper funciton that loads model data for visualisation.

Commit:
Create binary_threshold_vs_metric_cli wrapper funciton that loads model data for visualisation.

Commit:
Create binary_threshold_vs_metric_cli wrapper funciton that loads model data for visualisation.

Commit:
Create roc_curves_cli wrapper function that loads model data for visualisation.

Commit:
Create roc_curves_cli wrapper function that loads model data for visualisation.

Commit:
Create roc_curves_from_test_statistics_cli wrapper function that loads model data for visualisation.

Commit:
Create roc_curves_from_test_statistics_cli wrapper function that loads model data for visualisation.

Commit:
Create calibration_1_vs_all_cli wrapper function that loads model data for visualisation.

Commit:
Create calibration_1_vs_all_cli wrapper function that loads model data for visualisation.

Commit:
Create calibration_multiclass_cli and frequency_vs_f1_cli wrapper function that loads model data for visualisation.

Commit:
Create calibration_multiclass_cli and frequency_vs_f1_cli wrapper function that loads model data for visualisation.

Commit:
Create confusion_matrix_cli and frequency_vs_f1_cli wrapper function that loads model data for visualisation.

Commit:
Create confusion_matrix_cli and frequency_vs_f1_cli wrapper function that loads model data for visualisation.

Commit:
Create supported visualisation config and use it invoke visualisations.

Commit:
Create supported visualisation config and use it invoke visualisations.

Commit:
Remove comments

Commit:
Remove comments

Commit:
New line

Commit:
Merge remote-tracking branch 'origin/feature/visualisation_api' into feature/visualisation_api

# Conflicts:
#	docs/404.html
#	docs/api/index.html
#	docs/developer_guide/index.html
#	docs/examples/index.html
#	docs/faq/index.html
#	docs/getting_started/index.html
#	docs/user_guide/index.html
